src/oid4vp/holder.rs

```diff
@@ -0,0 +1,410 @@
+    wallet::Wallet as OID4VPWallet,
+};
+use serde_json::Value;
+use ssi::claims::jws::JwsSigner;
+use ssi::claims::vc;
+use ssi::claims::vc::syntax::IdOr;
+use ssi::dids::DIDURLBuf;
+use ssi::json_ld::iref::UriBuf;
+use ssi::JWK;
+use uniffi::deps::anyhow;
+
+/// A Holder is an entity that possesses one or more Verifiable Credentials.
+/// The Holder is typically the subject of the credentials, but not always.
+/// The Holder has the ability to generate Verifiable Presentations from
+/// these credentials and share them with Verifiers.
+#[derive(Debug, uniffi::Object)]
+pub struct Holder {
+    /// An atomic reference to the VDC collection.
+    pub(crate) vdc_collection: Arc<VdcCollection>,
+
+    /// Metadata about the holder.
+    pub(crate) metadata: WalletMetadata,
+
+    /// HTTP Request Client
+    pub(crate) client: oid4vp::core::util::ReqwestClient,
+
+    /// A list of trusted DIDs.
+    pub(crate) trusted_dids: Vec<String>,
+
+    /// A request signer callback interface to handle
+    /// cryptographic signing with native libraries.
+    pub(crate) request_signer: Arc<dyn RequestSignerInterface>,
+}
+
+#[uniffi::export]
+impl Holder {
+    #[uniffi::constructor]
+    pub async fn new(
+        vdc_collection: Arc<VdcCollection>,
+        trusted_dids: Option<Vec<String>>,
+        request_signer: Arc<dyn RequestSignerInterface>,
+    ) -> Result<Arc<Self>, Error> {
+        let client = oid4vp::core::util::ReqwestClient::new()
+            .map_err(|e| Error::HttpClientInitialization(e.to_string()))?;
+
+        Ok(Arc::new(Self {
+            client,
+            vdc_collection,
+            metadata: WalletMetadata::openid4vp_scheme_static(),
+            trusted_dids: trusted_dids.unwrap_or_default(),
+            request_signer,
+        }))
+    }
+
+    /// Handle an OID4VP authorization request provided as a URL.
+    ///
+    /// This method will validate and process the request, returning a
+    /// redirect URL with the encoded verifiable presentation token,
+    /// if the presentation exchange was successful.
+    ///
+    /// If the request is invalid or cannot be processed, an error will be returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - The URL containing the OID4VP authorization request.
+    /// * `callback` - A callback interface to retrieve credentials.
+    /// * `provided_credentials` - An optional list of credentials to use for the presentation,
+    ///     this is used instead of search for the credentials in the VDC collection.
+    ///
+    /// # Returns
+    ///
+    /// An optional URL containing the OID4VP response.
+    ///
+    /// # Errors
+    ///
+    /// * If the request is invalid;
+    /// * If the response mode is not supported;
+    /// * If the response submission fails.
+    pub async fn handle_oid4vp_request(
+        &self,
+        url: Url,
+        callback: &Arc<dyn CredentialCallbackInterface>,
+        provided_credentials: Option<Vec<Arc<Credential>>>,
```

Comment:
NOTE: this argument is provided to bypass the VDC credential lookup, but we may wish to move this to the constructor method of the `Holder::new`, which will be used to check internally.

A new constructor method might be used, e.g., `Holder::new_with_credentials`, etc.